src/dmqproto/client/request/internal/Consume.d

```diff
@@ -12,6 +12,711 @@
+                        this.fiber.suspend(token, this);
+                        this.fiber_suspended = fiber_suspended.No;
+                    }
+
+                    this.passRecordToUser(
+                        this.outer.conn.message_parser.getArray
+                        !(Const!(void))(this.remaining_batch)
+                    );
+
+                    if (!this.remaining_batch.length)
+                        break;
+
+                    if (yield_count >= 10) // yield every 10th record
+                    {
+                        yield_count = 0;
+                        this.outer.request_event_dispatcher.yield(this.fiber);
+                    }
+                }
+
+                this.remaining_batch = null;
+                (*this.batch_buffer).length = 0;
+
+                if (this.stopped)
+                    break;
+
+                this.outer.request_event_dispatcher.send(
+                    this.fiber,
+                    (conn.Payload payload)
+                    {
+                        payload.addConstant(MessageType_v3.Continue);
+                    }
+                );
+            }
+
+            this.outer.request_event_dispatcher.signal(this.outer.conn,
+                Controller.terminate_signal);
+        }
+
+        /***********************************************************************
+
+            Suspends the fiber to be resumed by `addRecords` or `stop`.
+
+            Returns:
+                true if the fiber was resumed by `addRecords` or false if
+                resumed by `stop`.
+
+        ***********************************************************************/
+
+        private bool waitForRecords ( )
+        {
+            this.fiber_suspended = fiber_suspended.WaitingForRecords;
+            this.fiber.suspend(token, this);
+            this.fiber_suspended = fiber_suspended.No;
+            return !this.stopped;
+        }
+
+        /***********************************************************************
+
+            Calls the user notifier to pass `record` to the user. Handles a
+            request state change (i.e. stopping the request) if the user uses
+            the controller in the notifier.
+
+            Params:
+                record = the record to pass to the user
+
+        ***********************************************************************/
+
+        private void passRecordToUser ( in void[] record )
+        {
+            bool initially_stopped =
+                this.outer.context.shared_working.suspendable_control.stopped;
+
+            Notification notification;
+            notification.received = RequestDataInfo(this.outer.context.request_id, record);
+            Consume.notify(this.outer.context.user_params, notification);
+
+            if (!initially_stopped &&
+                this.outer.context.shared_working.suspendable_control.stopped
+            )
+                this.outer.sendStopMessage(this.fiber);
```

Comment:
Hmm why do this here and not in the controller fiber?

Reply:
Good idea.